File: go/dtn/envelope.go
// SPDX-License-Identifier: MIT

package dtn

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"io"
	"time"

	"github.com/google/uuid"

	"github.com/bhengubv/aether-protocol/go/models"
)

// Binary DTN envelope serialization — the cross-language wire format for the
// three DTN packet bodies (bundle / custody-ack / delivery-receipt) carried in
// MeshPacket.Payload. Conventions mirror the packet serializer (go/protocol):
// all multi-byte integers are little-endian; the 16-byte bundle id is the UUID
// in RFC-4122 big-endian order (uuid.MarshalBinary); strings are uint16-LE
// length-prefixed UTF-8; the encrypted payload is int32-LE length-prefixed raw
// bytes. Every envelope begins with a single format-version byte so the format
// can evolve (e.g. a future move of routing metadata inside the encrypted body)
// without a flag-day — a reader rejects any unknown version.
//
// Cleartext routing fields are laid out first and the opaque encrypted_payload
// last, so a later version can encrypt sender/recipient with no field-shuffle.

// DtnEnvelopeVersion is the format-version byte at offset 0 of every envelope.
const DtnEnvelopeVersion byte = 0x01

// maxEnvelopePayload bounds the encrypted_payload length on read so a hostile
// i32 length cannot trigger a huge allocation. Mirrors AETHERNET_MAX_PAYLOAD_LEN.
const maxEnvelopePayload = 16 * 1024 * 1024

// SerializeBundle encodes a DtnBundle into its binary envelope.
func SerializeBundle(b *models.DtnBundle) ([]byte, error) {
	if b == nil {
		return nil, fmt.Errorf("dtn: bundle must not be nil")
	}
	idBytes, err := marshalUUID(b.ID)
	if err != nil {
		return nil, err
	}
	buf := new(bytes.Buffer)
	buf.WriteByte(DtnEnvelopeVersion)
	buf.Write(idBytes)
	buf.WriteByte(byte(b.Priority))
	buf.WriteByte(byte(b.Status))
	writeI32(buf, b.CopyCount)
	writeI32(buf, b.MaxCopies)
	writeI32(buf, b.HopCount)
	writeI64(buf, b.CreatedAt.UnixMilli())
	writeI64(buf, b.ExpiresAt.UnixMilli())
	if err := writeStr(buf, b.SenderUhid); err != nil {
		return nil, err
	}
	if err := writeStr(buf, b.RecipientUhid); err != nil {
		return nil, err
	}
	if err := writeStr(buf, b.SenderGeohash); err != nil {
		return nil, err
	}
	if err := writeStr(buf, b.RecipientLastGeohash); err != nil {
		return nil, err
	}
	if err := writeBytes32(buf, b.EncryptedPayload); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// DeserializeBundle decodes a DtnBundle from its binary envelope.
func DeserializeBundle(data []byte) (*models.DtnBundle, error) {
	r := bytes.NewReader(data)
	if err := expectVersion(r); err != nil {
		return nil, err
	}
	id, err := readUUID(r)
	if err != nil {
		return nil, err
	}
	priority, err := r.ReadByte()
	if err != nil {
		return nil, fmt.Errorf("dtn: read priority: %w", err)
	}
	if priority > byte(models.DtnPrioritySos) {
		return nil, fmt.Errorf("dtn: invalid priority %d", priority)
	}
	status, err := r.ReadByte()
	if err != nil {
		return nil, fmt.Errorf("dtn: read status: %w", err)
	}
	if status > byte(models.DtnStatusFailed) {
		return nil, fmt.Errorf("dtn: invalid status %d", status)
	}
	copyCount, err := readI32(r)
	if err != nil {
		return nil, err
	}
	maxCopies, err := readI32(r)
	if err != nil {
		return nil, err
	}
	hopCount, err := readI32(r)
	if err != nil {
		return nil, err
	}
	createdMs, err := readI64(r)
	if err != nil {
		return nil, err
	}
	expiresMs, err := readI64(r)
	if err != nil {
		return nil, err
	}
	senderUhid, err := readStr(r)
	if err != nil {
		return nil, err
	}
	recipientUhid, err := readStr(r)
	if err != nil {
		return nil, err
	}
	senderGeohash, err := readStr(r)
	if err != nil {
		return nil, err
	}
	recipientLastGeohash, err := readStr(r)
	if err != nil {
		return nil, err
	}
	payload, err := readBytes32(r)
	if err != nil {
		return nil, err
	}
	return &models.DtnBundle{
		ID:                   id,
		SenderUhid:           senderUhid,
		RecipientUhid:        recipientUhid,
		EncryptedPayload:     payload,
		Priority:             models.DtnPriority(priority),
		Status:               models.DtnStatus(status),
		CopyCount:            copyCount,
		MaxCopies:            maxCopies,
		SenderGeohash:        senderGeohash,
		RecipientLastGeohash: recipientLastGeohash,
		HopCount:             hopCount,
		CreatedAt:            time.UnixMilli(createdMs),
		ExpiresAt:            time.UnixMilli(expiresMs),
	}, nil
}

// SerializeCustodyAck encodes a custody-ack: version | bundle_id(16 BE) | accepted(u8).
func SerializeCustodyAck(bundleID string, accepted bool) ([]byte, error) {
	idBytes, err := marshalUUID(bundleID)
	if err != nil {
		return nil, err
	}
	buf := new(bytes.Buffer)
	buf.WriteByte(DtnEnvelopeVersion)
	buf.Write(idBytes)
	if accepted {
		buf.WriteByte(0x01)
	} else {
		buf.WriteByte(0x00)
	}
	return buf.Bytes(), nil
}

// DeserializeCustodyAck decodes a custody-ack.
func DeserializeCustodyAck(data []byte) (bundleID string, accepted bool, err error) {
	r := bytes.NewReader(data)
	if err = expectVersion(r); err != nil {
		return "", false, err
	}
	id, err := readUUID(r)
	if err != nil {
		return "", false, err
	}
	acc, err := r.ReadByte()
	if err != nil {
		return "", false, fmt.Errorf("dtn: read accepted: %w", err)
	}
	return id, acc != 0, nil
}

// SerializeDeliveryReceipt encodes a delivery-receipt.
func SerializeDeliveryReceipt(bundleID, recipientUhid string, totalHops, totalCustodyTransfers int32, deliveredAtMs int64) ([]byte, error) {
	idBytes, err := marshalUUID(bundleID)
	if err != nil {
		return nil, err
	}
	buf := new(bytes.Buffer)
	buf.WriteByte(DtnEnvelopeVersion)
	buf.Write(idBytes)
	if err := writeStr(buf, recipientUhid); err != nil {
		return nil, err
	}
	writeI32(buf, totalHops)
	writeI32(buf, totalCustodyTransfers)
	writeI64(buf, deliveredAtMs)
	return buf.Bytes(), nil
}

// DeserializeDeliveryReceipt decodes a delivery-receipt.
func DeserializeDeliveryReceipt(data []byte) (bundleID, recipientUhid string, totalHops, totalCustodyTransfers int32, deliveredAtMs int64, err error) {
	r := bytes.NewReader(data)
	if err = expectVersion(r); err != nil {
		return
	}
	bundleID, err = readUUID(r)
	if err != nil {
		return
	}
	recipientUhid, err = readStr(r)
	if err != nil {
		return
	}
	totalHops, err = readI32(r)
	if err != nil {
		return
	}
	totalCustodyTransfers, err = readI32(r)
	if err != nil {
		return
	}
	deliveredAtMs, err = readI64(r)
	return
}

// ---- low-level helpers (mirror go/protocol/serializer.go conventions) -------

func marshalUUID(s string) ([]byte, error) {
	id, err := uuid.Parse(s)
	if err != nil {
		return nil, fmt.Errorf("dtn: invalid uuid %q: %w", s, err)
	}
	b, err := id.MarshalBinary() // RFC-4122 big-endian, 16 bytes
	if err != nil {
		return nil, fmt.Errorf("dtn: marshal uuid: %w", err)
	}
	return b, nil
}

func readUUID(r io.Reader) (string, error) {
	b := make([]byte, 16)
	if _, err := io.ReadFull(r, b); err != nil {
		return "", fmt.Errorf("dtn: read uuid: %w", err)
	}
	var id uuid.UUID
	if err := id.UnmarshalBinary(b); err != nil {
		return "", fmt.Errorf("dtn: unmarshal uuid: %w", err)
	}
	return id.String(), nil
}

func expectVersion(r *bytes.Reader) error {
	v, err := r.ReadByte()
	if err != nil {
		return fmt.Errorf("dtn: read version: %w", err)
	}
	if v != DtnEnvelopeVersion {
		return fmt.Errorf("dtn: unsupported envelope version 0x%02x", v)
	}
	return nil
}

func writeI32(buf *bytes.Buffer, v int32) { _ = binary.Write(buf, binary.LittleEndian, v) }
func writeI64(buf *bytes.Buffer, v int64) { _ = binary.Write(buf, binary.LittleEndian, v) }

func writeStr(buf *bytes.Buffer, s string) error {
	b := []byte(s)
	if len(b) > 65535 {
		return fmt.Errorf("dtn: string too long: %d bytes exceeds 65535", len(b))
	}
	_ = binary.Write(buf, binary.LittleEndian, uint16(len(b)))
	buf.Write(b)
	return nil
}

func writeBytes32(buf *bytes.Buffer, data []byte) error {
	if len(data) > maxEnvelopePayload {
		return fmt.Errorf("dtn: payload too large: %d bytes exceeds %d", len(data), maxEnvelopePayload)
	}
	_ = binary.Write(buf, binary.LittleEndian, int32(len(data)))
	buf.Write(data)
	return nil
}

func readI32(r io.Reader) (int32, error) {
	var v int32
	err := binary.Read(r, binary.LittleEndian, &v)
	return v, err
}

func readI64(r io.Reader) (int64, error) {
	var v int64
	err := binary.Read(r, binary.LittleEndian, &v)
	return v, err
}

func readStr(r io.Reader) (string, error) {
	var n uint16
	if err := binary.Read(r, binary.LittleEndian, &n); err != nil {
		return "", err
	}
	if n == 0 {
		return "", nil
	}
	b := make([]byte, n)
	if _, err := io.ReadFull(r, b); err != nil {
		return "", err
	}
	return string(b), nil
}

func readBytes32(r io.Reader) ([]byte, error) {
	var n int32
	if err := binary.Read(r, binary.LittleEndian, &n); err != nil {
		return nil, err
	}
	if n < 0 {
		return nil, fmt.Errorf("dtn: negative payload length: %d", n)
	}
	if int64(n) > maxEnvelopePayload {
		return nil, fmt.Errorf("dtn: payload length %d exceeds %d", n, maxEnvelopePayload)
	}
	if n == 0 {
		return []byte{}, nil
	}
	b := make([]byte, n)
	if _, err := io.ReadFull(r, b); err != nil {
		return nil, err
	}
	return b, nil
}

File: rust/src/dtn/envelope.rs
// SPDX-License-Identifier: MIT

//! Canonical binary DTN envelope (wire format v1). Byte-identical across all
//! eight AetherNet SDKs; the Go encoder (`go/cmd/dtnfixturegen`) is the oracle
//! and `fixtures/dtn/expected/*.bin` pins the bytes.
//!
//! Layout — every multi-byte integer is LITTLE-ENDIAN, except the 16-byte
//! bundle id which is RFC-4122 BIG-ENDIAN (`Uuid::as_bytes`, mirroring the
//! packet serializer; never the legacy .NET mixed-endian Guid layout).
//! Cleartext routing fields come first and the opaque `encrypted_payload` is
//! last, so the future T1 privacy bump can move sender/recipient into the
//! ciphertext without a re-layout.
//!
//! The Rust [`DtnBundle`] stores `created_at`/`expires_at` as Unix **seconds**;
//! the wire carries milliseconds, so we multiply/divide by 1000 (every fixture
//! timestamp is whole-second-aligned, so this round-trips exactly).

use uuid::Uuid;

use crate::models::{BundlePriority, BundleStatus, DtnBundle};

const VERSION: u8 = 0x01;
const MAX_PAYLOAD: usize = 16 * 1024 * 1024; // AETHERNET_MAX_PAYLOAD_LEN

// ─────────────────────────────── DtnBundle ───────────────────────────────

pub fn serialize_bundle(b: &DtnBundle) -> Vec<u8> {
    let mut out = Vec::with_capacity(64 + b.encrypted_payload.len());
    out.push(VERSION);
    out.extend_from_slice(b.id.as_bytes()); // 16 bytes, RFC-4122 big-endian
    out.push(b.priority.as_u8());
    out.push(b.status.as_u8());
    write_i32(&mut out, b.copy_count);
    write_i32(&mut out, b.max_copies);
    write_i32(&mut out, b.hop_count);
    write_i64(&mut out, (b.created_at as i64) * 1000);
    write_i64(&mut out, (b.expires_at as i64) * 1000);
    write_str(&mut out, &b.sender_uhid);
    write_str(&mut out, &b.recipient_uhid);
    write_str(&mut out, b.sender_geohash.as_deref().unwrap_or(""));
    write_str(&mut out, b.recipient_last_geohash.as_deref().unwrap_or(""));
    write_bytes32(&mut out, &b.encrypted_payload);
    out
}

pub fn deserialize_bundle(data: &[u8]) -> Option<DtnBundle> {
    let mut r = Reader::new(data);
    r.version()?;
    let id = r.uuid()?;
    let priority = r.u8()?;
    if priority > 3 {
        return None;
    }
    let status = r.u8()?;
    if status > 4 {
        return None;
    }
    let copy_count = r.i32()?;
    let max_copies = r.i32()?;
    let hop_count = r.i32()?;
    let created_at_ms = r.i64()?;
    let expires_at_ms = r.i64()?;
    let sender_uhid = r.string()?;
    let recipient_uhid = r.string()?;
    let sender_geohash = r.string()?;
    let recipient_last_geohash = r.string()?;
    let encrypted_payload = r.bytes32()?;
    Some(DtnBundle {
        id,
        sender_uhid,
        recipient_uhid,
        encrypted_payload,
        priority: BundlePriority::from_u8(priority),
        status: BundleStatus::from_u8(status),
        copy_count,
        max_copies,
        sender_geohash: empty_to_none(sender_geohash),
        recipient_last_geohash: empty_to_none(recipient_last_geohash),
        hop_count,
        created_at: (created_at_ms / 1000) as u64,
        expires_at: (expires_at_ms / 1000) as u64,
    })
}

// ─────────────────────── CustodyAck (18 bytes fixed) ──────────────────────

pub fn serialize_custody_ack(bundle_id: &Uuid, accepted: bool) -> Vec<u8> {
    let mut out = Vec::with_capacity(18);
    out.push(VERSION);
    out.extend_from_slice(bundle_id.as_bytes());
    out.push(if accepted { 1 } else { 0 });
    out
}

pub fn deserialize_custody_ack(data: &[u8]) -> Option<(Uuid, bool)> {
    let mut r = Reader::new(data);
    r.version()?;
    let id = r.uuid()?;
    let accepted = r.u8()? != 0;
    Some((id, accepted))
}

// ───────────────────────────── DeliveryReceipt ────────────────────────────

pub fn serialize_delivery_receipt(
    bundle_id: &Uuid,
    recipient_uhid: &str,
    total_hops: i32,
    total_custody_transfers: i32,
    delivered_at_ms: i64,
) -> Vec<u8> {
    let mut out = Vec::with_capacity(64);
    out.push(VERSION);
    out.extend_from_slice(bundle_id.as_bytes());
    write_str(&mut out, recipient_uhid);
    write_i32(&mut out, total_hops);
    write_i32(&mut out, total_custody_transfers);
    write_i64(&mut out, delivered_at_ms);
    out
}

/// Returns `(bundle_id, recipient_uhid, total_hops, total_custody_transfers, delivered_at_ms)`.
pub fn deserialize_delivery_receipt(data: &[u8]) -> Option<(Uuid, String, i32, i32, i64)> {
    let mut r = Reader::new(data);
    r.version()?;
    let id = r.uuid()?;
    let recipient = r.string()?;
    let total_hops = r.i32()?;
    let total_custody = r.i32()?;
    let delivered_at_ms = r.i64()?;
    Some((id, recipient, total_hops, total_custody, delivered_at_ms))
}

// ─────────────────────────────── primitives ───────────────────────────────

fn empty_to_none(s: String) -> Option<String> {
    if s.is_empty() {
        None
    } else {
        Some(s)
    }
}

fn write_i32(out: &mut Vec<u8>, v: i32) {
    out.extend_from_slice(&v.to_le_bytes());
}

fn write_i64(out: &mut Vec<u8>, v: i64) {
    out.extend_from_slice(&v.to_le_bytes());
}

fn write_str(out: &mut Vec<u8>, s: &str) {
    let bytes = s.as_bytes();
    let len = bytes.len().min(u16::MAX as usize);
    out.extend_from_slice(&(len as u16).to_le_bytes());
    out.extend_from_slice(&bytes[..len]);
}

fn write_bytes32(out: &mut Vec<u8>, b: &[u8]) {
    out.extend_from_slice(&(b.len() as i32).to_le_bytes());
    out.extend_from_slice(b);
}

struct Reader<'a> {
    data: &'a [u8],
    pos: usize,
}

impl<'a> Reader<'a> {
    fn new(data: &'a [u8]) -> Self {
        Self { data, pos: 0 }
    }

    fn version(&mut self) -> Option<()> {
        if self.u8()? == VERSION {
            Some(())
        } else {
            None
        }
    }

    fn u8(&mut self) -> Option<u8> {
        let b = *self.data.get(self.pos)?;
        self.pos += 1;
        Some(b)
    }

    fn take(&mut self, n: usize) -> Option<&'a [u8]> {
        let end = self.pos.checked_add(n)?;
        let slice = self.data.get(self.pos..end)?;
        self.pos = end;
        Some(slice)
    }

    fn uuid(&mut self) -> Option<Uuid> {
        let slice = self.take(16)?;
        let mut arr = [0u8; 16];
        arr.copy_from_slice(slice);
        Some(Uuid::from_bytes(arr))
    }

    fn i32(&mut self) -> Option<i32> {
        let slice = self.take(4)?;
        Some(i32::from_le_bytes(slice.try_into().ok()?))
    }

    fn i64(&mut self) -> Option<i64> {
        let slice = self.take(8)?;
        Some(i64::from_le_bytes(slice.try_into().ok()?))
    }

    fn u16(&mut self) -> Option<u16> {
        let slice = self.take(2)?;
        Some(u16::from_le_bytes(slice.try_into().ok()?))
    }

    fn string(&mut self) -> Option<String> {
        let n = self.u16()? as usize;
        let slice = self.take(n)?;
        String::from_utf8(slice.to_vec()).ok()
    }

    fn bytes32(&mut self) -> Option<Vec<u8>> {
        let n = self.i32()?;
        if n < 0 || n as usize > MAX_PAYLOAD {
            return None;
        }
        let slice = self.take(n as usize)?;
        Some(slice.to_vec())
    }
}
